### PR TITLE
Add Discord rich presence

### PR DESCRIPTION
**Pull request by heinrich5991**
Continued the work of @edg-l. Supersedes #3472.

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/engine/client/discord.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	void ClearGameInfo()
+	{
+		m_pActivityManager->clear_activity(m_pActivityManager, 0, 0);
+	}
+	void SetGameInfo(NETADDR ServerAddr, const char *pMapName)
+	{
+		DiscordActivity Activity;
+		mem_zero(&Activity, sizeof(DiscordActivity));
+		str_copy(Activity.assets.large_image, "ddnet_logo", sizeof(Activity.assets.large_image));
+		str_copy(Activity.assets.large_text, "DDNet logo", sizeof(Activity.assets.large_text));
```
Comment (heinrich5991) [inline src/engine/client/discord.cpp:49]: Do we need this, or is this automatic?

File: src/engine/client/discord.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		Activity.timestamps.start = time_timestamp(); // TODO: Should this be in UTC or in local time?
```
Comment (heinrich5991) [inline src/engine/client/discord.cpp:50]: Should this be in UTC or in local time?

File: src/engine/client/discord.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		str_copy(Activity.details, pMapName, sizeof(Activity.details));
+		str_copy(Activity.state, "Online", sizeof(Activity.state)); // TODO: Can we leave this empty instead?
```
Comment (heinrich5991) [inline src/engine/client/discord.cpp:52]: Can we leave this empty instead? Discord argues that we should only put information dense stuff there.

**Review by heinrich5991 — COMMENTED** [review]
Changes from the original pull request:

- Removed config variable.
- Removed the rich presence when not in game (see not much use of this, would need to distinguish between "in menus" and demos etc.

@edg-l How do I test this? When I start it, it automatically closes and opens discord:///library/752165779117441075/launch in the browser.

@Learath2 Can you investigate whether we're allowed to distribute Discord SDK? I couldn't find anything in that regard.

**Comment by Kaffeine** [comment]
I would say that there should be a single `IPresencePublisher` interface with two implementations — Steam and Discord (I don't think that we'll have use-cases for the Stubs).
Current `Steam()->` and the upcoming `Discord()->` can/should be replaced with code like:
```cpp
for(IPresencePublisher *publisher : GetPresencePublishers())
{
    publisher->Update();
    if(publisher->GetConnectAddress())
    {
        char aAddress[NETADDR_MAXSTRSIZE];
        net_addr_str(publisher->GetConnectAddress(), aAddress, sizeof(aAddress), true);
        Connect(aAddress);
        publisher->ClearConnectAddress();
    }
}
```

We can continue to have `CreateSteam()` and `CreateDiscord()` and now it will be OK for them to return `nullptr` instead of the stubs.

**Comment by heinrich5991** [comment]
> I would say that there should be a single `IPresencePublisher` interface with two implementations — Steam and Discord (I don't think that we'll have use-cases for the Stubs). Maybe we'll want to implement some GOG presence publisher as well.
> Current `Steam()->` and the upcoming `Discord()->` can/should be replaced with code like:

I considered this, but thought it would be overkill for just two of them.

**Comment by Kaffeine** [comment]
> I considered this, but thought it would be overkill for just two of them.

Yes, it is a very valid concern and I don't have a strong opinion here. It's just thoughts and "would" 😊.

We don't have many duplications now, as there is no `IDiscord::GetConnectAddress()` yet, but probably we'll have it one day.
```
if(Discord()->GetConnectAddress())
{
    char aAddress[NETADDR_MAXSTRSIZE];
    net_addr_str(Discord()->GetConnectAddress(), aAddress, sizeof(aAddress), true);
    Connect(aAddress);
    Discord()->ClearConnectAddress();
}
```

Pros:
- We can remove `IDiscord` in favor of `IPresencePublisher` (renamed from `ISteam`)
- We can remove the stubs class and simplify the `Create*()` methods. 
- We'll have no code duplication nor references to implementations in the client code

Cons:
- We'll need as some (simple) `GetPresencePublishers()` implementation
- We may end up with a few lines more code, as long as we don't join a game via Discord (no GetConnectAddress() implementation).

**Comment by Kaffeine** [comment]
What a coincide. Six years ago I prototyped RichPresence for Linux (DBus). It actually worked. I wired it to `Psi+` XMPP client, so the Jabber client showed currently playing the game and let the contact-list join the game (the game had DBus-activation).

I had an idea of GCRIS (`Game Client Remote Interfacing Specification` analog of [MPRIS](https://specifications.freedesktop.org/mpris-spec/latest/) and some XMPP extension (analog of [XEP-0118: User Tune](https://xmpp.org/extensions/xep-0118.html)).

https://github.com/Kaffeine/teeworlds-ddrace/commit/25a43c7d4a0622ea00d2d6f5820b442deafabca4

This might be an implementation of `IPresencePublisher`.

**Comment by Learath2** [comment]
> @Learath2 Can you investigate whether we're allowed to distribute Discord SDK? I couldn't find anything in that regard.

I did quite a bit of research, seems https://discord.com/developers/docs/legal is the only license. I honestly couldn't be sure whether we can or can't redistribute this nor could I find any place where I could send a support ticket to a legalperson.

Seems the older `discord-rpc` library is very clearly licensed with MIT and is what's used by other open source projects like dolphin, rpcs3, citra etc. Maybe we should look into that or doing the rpc on our own maybe?

I would also be fine with the same stub approach as the one we have for steamapi, tho to be completely honest with you, I'm not even sure that is legal. They made it soooo hard for open source projects to get integrations with these things. It's honestly disappointing to see not one company actually being friendly to the open source community they so heavily rely on.

File: src/engine/client/discord.cpp (hunk 1 above)
Comment (edg-l) [reply src/engine/client/discord.cpp:49]: The asset is not automatic, based on when i made this.

File: src/engine/client/discord.cpp (hunk 2 above)
Comment (edg-l) [reply src/engine/client/discord.cpp:50]: UTC, u can test the timestamp on discord dev web

File: src/engine/client/discord.cpp (hunk 3 above)
Comment (edg-l) [reply src/engine/client/discord.cpp:52]: I guess

**Comment by heinrich5991** [comment]
Okay, this is now ready for review. I tested it with and without Discord client open, seems like you have to have the Discord client open before you launch the game.

I had to set the `EDiscordCreateFlags::DiscordCreateFlags_NoRequireDiscord` flag so that DDNet would still launch without Discord open.

@Learath2 I guess since there's no thing explicitly telling us not to redistribute it, we can try it without a stub for now.

**Comment by heinrich5991** [comment]
image

**Comment by def-** [comment]
If someone doesn't want Discord to know about the game running, they have to set it in Discord or compile the client themselves, right? No config for this.

Is it possible to link to the server? (Probably most players don't have a `ddnet:` link handler set up anyway)

**Comment by heinrich5991** [comment]
> If someone doesn't want Discord to know about the game running, they have to set it in Discord or compile the client themselves, right? No config for this.

This is correct. Discord has a setting for this:

image

**Comment by heinrich5991** [comment]
> Is it possible to link to the server? (Probably most players don't have a `ddnet:` link handler set up anyway)

I'll quickly look into this.

**Comment by heinrich5991** [comment]
I didn't find a quick way to do this. I'd like to not implement it in this PR.

**Comment by edg-l** [comment]
I think it's possible but requires quite some work, agree better leave it for a future pr.

**Comment by edg-l** [comment]
Has this been tested on windows and macos?

**Comment by def-** [comment]
Let's merge this and ask players to test in release candidate?

**Comment by Learath2** [comment]
Let me test it on macOS atleast, will let you know tonight

**Comment by heinrich5991** [comment]
I think we should at least try compiling it on Windows and macOS, it probably won't. I neither have a Windows nor a macOS machine ready.

**Comment by def-** [comment]
I'll do a test build of it now.

**Comment by def-** [comment]
Oh, it's disabled by default. Shouldn't we enable this flag?

**Comment by heinrich5991** [comment]
Enable it on CI? Yes, I guess.

I wouldn't want to enable it by default otherwise, it's a non-free dependency.

**Comment by def-** [comment]
I thought the idea is that we ship this in the official client, or am I wrong?

**Comment by Learath2** [comment]
I guess he means as in a default configuration from cmake, just like we don't include libsteam

**Comment by def-** [comment]
So how can I build this? Do I need to get the discord libs from somewhere? Or just set DISCORD enabled in cmake for the official builds?

**Comment by def-** [comment]
Ok, this is not building on Linux and Windows.

On  Linux:
```
ddnet-libs/discord/linux/lib64/discord_game_sdk.so: undefined reference to `exp2f@GLIBC_2.27'
ddnet-libs/discord/linux/lib64/discord_game_sdk.so: undefined reference to `powf@GLIBC_2.27'
```

On Windows:
```
CMake Error in CMakeLists.txt:
  IMPORTED_IMPLIB not set for imported target "discord-shared" configuration
  "Release".
```

**Comment by def-** [comment]
Linux x86 actually fails differently:
```
CMake Error at CMakeLists.txt:492 (message):
  You must install the Discord SDK to compile the DDNet client with Discord
  support
```

**Comment by def-** [comment]
Can we build the Discord SDK ourselves or do we have to use the shared libs they provide?

**Comment by heinrich5991** [comment]
> Can we build the Discord SDK ourselves or do we have to use the shared libs they provide?

They don't provide source code, we have to use the shared libs they provide.

**Comment by def-** [comment]
This means we have to increase the glibc version of our Linux releases I guess. glibc 2.27 sounds quite new, many would fail here: https://gist.github.com/wagenet/35adca1a032cec2999d47b6c40aa45b1

Should we not bundle the Discord presence for Linux?

**Comment by Learath2** [comment]
As I said we could look into discord-rpc instead which has source code and
is MIT licensed

**Comment by heinrich5991** [comment]
> As I said we could look into discord-rpc instead which has source code and is MIT licensed

It seems that Discord RPC has already been deprecated: https://github.com/discord/discord-rpc/blob/963aa9f3e5ce81a4682c6ca3d136cddda614db33/README.md#deprecation-notice, https://discord.com/developers/docs/rich-presence/how-to.

**Comment by Learath2** [comment]
Yes it's deprecated but I believe they said in couple places that it'll keep working as the GameSDK itself relies on that interface. Barring that, could we maybe look into dynamically loading the library if glibc version is > what the gamesdk is linked to?

**Comment by heinrich5991** [comment]
> Or now that I think about it, should we maybe statically link glibc? It could help avoid a lot of linux compatibility issues.

We can't force the SDK to statically link libc (AFAIK).

**Comment by def-** [comment]
So for now let's only do this on Windows and Mac for now, not on Linux. There you can use your distribution's versions of compile yourself if you want Discord rich presence. Ok?

@heinrich5991 Any idea how to fix this?
```
CMake Error in CMakeLists.txt:
  IMPORTED_IMPLIB not set for imported target "discord-shared" configuration
  "Release".
```

**Comment by heinrich5991** [comment]
Haven't looked at this yet, but maybe @edg-l knows something, he wrote that code.

**Comment by def-** [comment]
bors r+
Rebased

**Pull request merged**